homeassistant/components/application_credentials/__init__.py

```diff
@@ -0,0 +1,215 @@
+        """Return a new updated data object."""
+        raise ValueError("Updates not supported")
+
+    async def async_delete_item(self, item_id: str) -> None:
+        """Delete item, verifying credential is not in use."""
+        if item_id not in self.data:
+            raise collection.ItemNotFound(item_id)
+
+        # Cannot delete a credential currently in use by a ConfigEntry
+        current = self.data[item_id]
+        entries = self.hass.config_entries.async_entries(current[CONF_DOMAIN])
+        for entry in entries:
+            if entry.data.get("auth_implementation") == item_id:
+                raise ValueError("Cannot delete credential in use by an integration")
+
+        await super().async_delete_item(item_id)
+
+    def async_client_credentials(self, domain: str) -> dict[str, ClientCredential]:
+        """Return ClientCredentials in storage for the specified domain."""
+        credentials = {}
+        for item in self.async_items():
+            if item[CONF_DOMAIN] != domain:
+                continue
+            auth_domain = item.get(CONF_AUTH_IMPL, item[CONF_ID])
+            credentials[auth_domain] = ClientCredential(
+                item[CONF_CLIENT_ID], item[CONF_CLIENT_SECRET]
+            )
+        return credentials
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Application Credentials."""
+    hass.data[DOMAIN] = {}
+
+    id_manager = collection.IDManager()
+    storage_collection = ApplicationCredentialsStorageCollection(
+        Store(hass, STORAGE_VERSION, STORAGE_KEY),
+        logging.getLogger(f"{__name__}.storage_collection"),
+        id_manager,
+    )
+    await storage_collection.async_load()
+    hass.data[DOMAIN][DATA_STORAGE] = storage_collection
+
+    collection.StorageCollectionWebsocket(
+        storage_collection, DOMAIN, DOMAIN, CREATE_FIELDS, UPDATE_FIELDS
+    ).async_setup(hass)
+
+    config_entry_oauth2_flow.async_register_local_apis(hass)
+    config_entry_oauth2_flow.async_add_implementation_provider(
+        hass, DOMAIN, _async_provide_implementation
+    )
+
+    return True
+
+
+async def async_import_client_credential(
+    hass: HomeAssistant, domain: str, auth_domain: str, credential: ClientCredential
+) -> None:
+    """Import an existing credential from configuration.yaml.
+
+    The imported credential auth domain should match the existing auth implementation domain used
+    when creating the LocalOAuth2Implementation, which is also persisted in the ConfigEntry.
+    """
+    if DOMAIN not in hass.data:
+        raise ValueError("Integration 'application_credentials' not setup")
+    storage_collection = hass.data[DOMAIN][DATA_STORAGE]
+    item = {
+        CONF_DOMAIN: domain,
+        CONF_CLIENT_ID: credential.client_id,
+        CONF_CLIENT_SECRET: credential.client_secret,
+        CONF_AUTH_IMPL: auth_domain,
+    }
+    await storage_collection.async_create_item(item)
```

Comment:
Shouldn't we deduplicate to avoid importing a credential multiple times?

Reply:
Done. I made a call to ignore existing entries, though this means that a changed client secret would NOT be imported (e.g you'd need to go remove from the UI if you want to change the value). This seemed nice and simple, but i'm still calibrating my judgement w.r.t. trade-offs for end users of the project so can also make this overwrite if that is what we'd prefer.